botblox_config/data_manager/tagvlan.py

```diff
@@ -0,0 +1,567 @@
+            return vlan
+        else:
+            return self._vlans[vlan_id]
+
+    def remove_vlan(self, vlan_id: int) -> None:
+        self._switch.check_feature(SwitchFeature.TAGGED_VLAN)
+        self._switch.check_feature(SwitchFeature.VLAN_TABLE)
+
+        if vlan_id in self._vlans:
+            del self._vlans[vlan_id]
+
+    def add_vlan_member(self, vlan: Union[int, VLAN], port: Port) -> None:
+        self._switch.check_feature(SwitchFeature.TAGGED_VLAN)
+        self._switch.check_feature(SwitchFeature.VLAN_TABLE)
+        if isinstance(vlan, int):
+            if vlan not in self._vlans:
+                vlan = self.add_vlan(vlan)
+            else:
+                vlan = self._vlans[vlan]
+        if port not in vlan.members:
+            vlan.members.append(port)
+
+    def remove_vlan_member(self, vlan: Union[int, VLAN], port: Port) -> None:
+        self._switch.check_feature(SwitchFeature.TAGGED_VLAN)
+        self._switch.check_feature(SwitchFeature.VLAN_TABLE)
+        if isinstance(vlan, int):
+            if vlan not in self._vlans:
+                vlan = self.add_vlan(vlan)
+        if port in vlan.members:
+            vlan.members.remove(port)
+
+    def set_port_vlan_header_action(self, vlan: int, port: Port, action: VLANHeaderAction) -> None:
+        port_config = self._port_configs[self._ports().index(port)]
+        if port_config.per_vlan_header_action is None:
+            port_config.per_vlan_header_action = dict()
+        port_config.per_vlan_header_action[vlan] = action
+
+    def reset(self) -> None:
+        self._reset = True
+
+    def _check_supports_per_port_vlan_mode(self, mode: VLANMode) -> None:
+        if mode == VLANMode.DISABLED:
+            self._switch.check_feature(SwitchFeature.PER_PORT_VLAN_MODE_DISABLE)
+        elif mode == VLANMode.OPTIONAL:
+            self._switch.check_feature(SwitchFeature.PER_PORT_VLAN_MODE_OPTIONAL)
+        elif mode == VLANMode.ENABLED:
+            self._switch.check_feature(SwitchFeature.PER_PORT_VLAN_MODE_ENABLE)
+        elif mode == VLANMode.STRICT:
+            self._switch.check_feature(SwitchFeature.PER_PORT_VLAN_MODE_STRICT)
+
+    def _check_supports_vlan_mode(self, mode: VLANMode) -> None:
+        if mode == VLANMode.OPTIONAL:
+            self._switch.check_feature(SwitchFeature.VLAN_MODE_OPTIONAL)
+        elif mode == VLANMode.ENABLED:
+            self._switch.check_feature(SwitchFeature.VLAN_MODE_ENABLE)
+        elif mode == VLANMode.STRICT:
+            self._switch.check_feature(SwitchFeature.VLAN_MODE_STRICT)
+
+    def apply_to_switch(self) -> None:
+        if isinstance(self._switch, IP175G):
+            self._apply_to_switch_ip175g()
+        else:
+            raise NotImplementedError()
+
+    def _apply_to_switch_ip175g(self) -> None:  # noqa: C901
+        assert isinstance(self._switch, IP175G)
+
+        if self._reset:
+            cast(PortListField, self._switch.fields["TAG_VLAN_EN"]).clear()
```

Comment:
the idea is good, but i don't know how to implement it

Reply:
do you have ideas? I mean, the casts are not needed per se because the objects practically are of the correct type, but I had to add them to tell PyCharm that it's okay... and maybe the casts would catch some errors if you change the type of a field but forget to change these lines of code